staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset_pruner.go

```diff
@@ -31,101 +32,119 @@ import (
+}
+
+func (a *ApplySet) pruneAll(ctx context.Context, dynamicClient dynamic.Interface, visitedUids sets.Set[types.UID], deleteOptions *ApplySetDeleteOptions) error {
+	allObjects, err := a.findAllObjectsToPrune(ctx, dynamicClient, visitedUids)
+	if err != nil {
+		return err
+	}
+
+	return a.deleteObjects(ctx, dynamicClient, allObjects, deleteOptions)
 }
 
-func (p *applySetPruner) prune(ctx context.Context, namespace string, mapping *meta.RESTMapping, applyset *ApplySet) error {
-	applysetLabelSelector := applyset.LabelSelectorForMembers()
+func (a *ApplySet) findObjectsToPrune(ctx context.Context, dynamicClient dynamic.Interface, visitedUids sets.Set[types.UID], namespace string, mapping *meta.RESTMapping) ([]PruneObject, error) {
+	applysetLabelSelector := a.LabelSelectorForMembers()
 
 	opt := metav1.ListOptions{
 		LabelSelector: applysetLabelSelector,
 	}
 
 	klog.V(2).Infof("listing objects for pruning; namespace=%q, resource=%v", namespace, mapping.Resource)
-	objects, err := p.dynamicClient.Resource(mapping.Resource).Namespace(namespace).List(ctx, opt)
+	objects, err := dynamicClient.Resource(mapping.Resource).Namespace(namespace).List(ctx, opt)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	var pruneObjects []PruneObject
 	for i := range objects.Items {
 		obj := &objects.Items[i]
 
 		uid := obj.GetUID()
-		if p.visitedUids.Has(uid) {
+		if visitedUids.Has(uid) {
 			continue
 		}
 		name := obj.GetName()
-		if p.dryRunStrategy != cmdutil.DryRunClient {
-			if err := p.delete(ctx, namespace, name, mapping); err != nil {
-				return fmt.Errorf("deleting %s/%s: %w", namespace, name, err)
+		pruneObjects = append(pruneObjects, PruneObject{
+			Name:      name,
+			Namespace: namespace,
+			Mapping:   mapping,
+			Object:    obj,
+		})
+
+	}
+	return pruneObjects, nil
+}
+
+func (a *ApplySet) deleteObjects(ctx context.Context, dynamicClient dynamic.Interface, pruneObjects []PruneObject, opt *ApplySetDeleteOptions) error {
+	for i := range pruneObjects {
+		pruneObject := &pruneObjects[i]
+
+		name := pruneObject.Name
+		namespace := pruneObject.Namespace
+		mapping := pruneObject.Mapping
+
+		if opt.DryRunStrategy != cmdutil.DryRunClient {
+			if err := a.delete(ctx, dynamicClient, namespace, name, mapping, opt); err != nil {
+				return fmt.Errorf("pruning %v %s/%s: %w", pruneObject.Mapping.GroupVersionKind.String(), namespace, name, err)
 			}
 		}
 
-		p.printer.PrintObj(obj, p.ioStreams.Out)
+		opt.Printer.PrintObj(pruneObject.Object, opt.IOStreams.Out)
+
 	}
 	return nil
 }
 
-func (p *applySetPruner) delete(ctx context.Context, namespace string, name string, mapping *meta.RESTMapping) error {
-	return runDelete(ctx, namespace, name, mapping, p.dynamicClient, p.cascadingStrategy, p.gracePeriod, p.dryRunStrategy == cmdutil.DryRunServer)
+func (a *ApplySet) delete(ctx context.Context, dynamicClient dynamic.Interface, namespace string, name string, mapping *meta.RESTMapping, o *ApplySetDeleteOptions) error {
```

Comment:
nit: would this make more sense as a method on `ApplySetDeleteOptions`? It uses three of its values, and doesn't interact with `ApplySet` at all. Or maybe just call `runDelete` directly from `deleteObjects` unless I'm missing some value in the indirection here.